src/plugins/gs-plugin-eos.c

```diff
@@ -234,8 +306,51 @@ gs_plugin_eos_refine_popular_app (GsPlugin *plugin,
 	    gs_app_get_metadata_item (app, "GnomeSoftware::ImageTile-css"))
 		return;
 
-	css = g_strdup_printf ("background-image: url('%s');", popular_bg);
-	gs_app_set_metadata (app, "GnomeSoftware::ImageTile-css", css);
+	url_basename = g_path_get_basename (popular_bg);
+
+	/* First take a hash of this URL and see if it is in our cache */
+	tile_cache_hash = g_compute_checksum_for_string (G_CHECKSUM_SHA256,
+	                                                 popular_bg,
+	                                                 -1);
+	cache_identifier = g_strdup_printf("%s-%s", tile_cache_hash, url_basename);
+	cache_filename = gs_utils_get_cache_filename ("eos-popular-app-thumbnails",
+	                                              cache_identifier,
+	                                              GS_UTILS_CACHE_FLAG_NONE,
+	                                              NULL);
+
+	/* Check to see if the file exists in the cache at the time we called this
+	 * function. If it does, then change the css so that the tile loads. Otherwise,
+	 * we'll need to asynchronously fetch the image from the server and write it
+	 * to the cache */
+	if (g_file_test (cache_filename, G_FILE_TEST_EXISTS)) {
+		gs_plugin_eos_update_tile_image_from_filename (app, cache_filename);
+		return;
+	}
+
+	soup_uri = soup_uri_new (popular_bg);
+	g_debug("Downloading thumbnail %s to %s", popular_bg, cache_filename);
+	if (!soup_uri || !SOUP_URI_VALID_FOR_HTTP (soup_uri)) {
+		g_debug ("Couldn't download %s, URL is not valid", popular_bg);
+		return;
+	}
+
+	/* XXX: Note that we might have multiple downloads in progress here. We
+	 * don't make any attempt to keep track of this. */
+	message = soup_message_new_from_uri (SOUP_METHOD_GET, soup_uri);
+	if (!message) {
+		g_debug ("Couldn't download %s, network not available", popular_bg);
+		return;
+	}
+
+	request_data = g_new0 (PopularBackgroundImageTileRequestData, 1);
+	request_data->app = app;
+	request_data->plugin = plugin;
+	request_data->cache_filename = g_steal_pointer (&cache_filename);
+
+	soup_session_queue_message (priv->soup_session,
+	                            g_steal_pointer(&message),
```

Comment:
Also my comment still stands... Does `soup_session_queue_message()` actually consume the passed-in reference? Otherwise it's just fine to pass the pointer as-is.